Objective-C/CBLDatabase.mm

```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
-        C4Transaction t(_c4db);
-        if (!t.begin())
-            return convertError(t.error(), outError);
+- (bool) resolveConflictInDocument: (NSString*)docID
+              withConflictResolver: (id<CBLConflictResolver>)conflictResolver
+                             error: (NSError**)outError {
+    while (true) {
+        CBLDocument* localDoc;
+        CBLDocument* remoteDoc;
+        CBL_LOCK(self) {
+            C4Transaction t(_c4db);
+            if (!t.begin())
+                return convertError(t.error(), outError);
+            
+            // Read local document:
+            localDoc = [[CBLDocument alloc] initWithDatabase: self documentID: docID
+                                              includeDeleted: YES error: outError];
+            if (!localDoc)
+                return false;
+            
+            // Read the conflicting remote revision:
+            remoteDoc = [[CBLDocument alloc] initWithDatabase: self documentID: docID
+                                               includeDeleted: YES error: outError];
+            if (!remoteDoc || ![remoteDoc selectConflictingRevision])
+                return false;
+            
+            if (!t.commit())
+                return convertError(t.error(), outError);
+        }
         
-        // Read local document:
-        CBLDocument* localDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                           documentID: docID
-                                                       includeDeleted: YES
-                                                                error: outError];
-        if (!localDoc)
-            return false;
+        conflictResolver = conflictResolver ?: [CBLConflictResolution default];
         
-        // Read the conflicting remote revision:
-        CBLDocument* remoteDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                            documentID: docID
-                                                        includeDeleted: YES
-                                                                 error: outError];
-        if (!remoteDoc || ![remoteDoc selectConflictingRevision])
-            return false;
+        Assert(localDoc == nil && remoteDoc == nil, @"Local and remote document shouldn't be empty \
+               at same time, when resolving conflict.");
+        localDoc = localDoc.isDeleted ? nil : localDoc;
+        remoteDoc = remoteDoc.isDeleted ? nil : remoteDoc;
+        CBLConflict* conflict = [[CBLConflict alloc] initWithLocalDocument: localDoc
+                                                            remoteDocument: remoteDoc];
         
         // Resolve conflict:
-        CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
-                   docID, localDoc.revID, remoteDoc.revID);
+        CBLDocument* resolvedDoc;
+        @try {
+            CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
```

Comment:
change mine to local and theirs to remote to make it consistency with the naming in CBLConflict.